torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,809 @@
+            # stream is complete.
+            p._full_param_padded.record_stream(current_stream)  # type: ignore[attr-defined]
+            # There may be external references to the Tensor Storage that we
+            # can't modify, such as references that are created by
+            # ctx.save_for_backward in the forward pass. Thus when we
+            # unshard parameters, we should reuse the original Tensor
+            # Storage object and unshard it in-place. For now, just resize
+            # the Storage to 0 to save memory.
+            free_storage_(p._full_param_padded)  # type: ignore[attr-defined]
+
+
+    @torch.no_grad()
+    def _use_param_local_shard(self, params: Optional[List[Parameter]] = None) -> None:
+        """Use local shard for a list of params."""
+        if params is None:
+            params = self.params
+        for p in params:
+            p.data = p._local_shard  # type: ignore[attr-defined]
+
+    def _assert_state(self, state: Union[TrainingState, List[TrainingState]]) -> None:
+        """Assert we are in the given state."""
+        # Since assert can be turned off and this error checking
+        # is really important, we use explicit error checking
+        # and raise a ValueError if needed.
+        if isinstance(state, TrainingState):
+            state = [state]
+        if self.training_state not in state:
+            msg = f"expected to be in states {state} but current state " f"is {self.training_state}"
+            # In case we are failing in the context of autograd hook, asserting
+            # may not generate useful msg. So, let's print it to be sure.
+            if self.rank == 0:
+                print(f"Asserting FSDP instance is: {self}")
+                print(f"ERROR: {msg}")
+                traceback.print_stack()
+            raise ValueError(msg)
+
+
+def _get_default_cuda_device(module: nn.Module) -> torch.device:
+    """Try to infer CUDA device from module parameters."""
+    try:
+        compute_device = next(module.parameters()).device
+        if compute_device.type == "cuda":
+            return compute_device
+    except StopIteration:
+        pass
+    # Fall back to current CUDA device
+    return torch.device("cuda")
+
+
+def free_storage_(data: torch.Tensor) -> None:
+    """Free underlying storage of a Tensor."""
+    if data.storage().size() > 0:
+        # Since we're modifying the Tensor's Storage directly, make sure the Tensor
+        # is the sole occupant of the Storage.
+        assert data.storage_offset() == 0
+        data.storage().resize_(0)  # type: ignore[attr-defined]
+
+
+@torch.no_grad()
+def alloc_storage_(data: torch.Tensor, size: torch.Size) -> None:
+    """Allocate storage for a tensor."""
+    if data.storage().size() == size.numel():  # no need to reallocate
+        return
+    assert data.storage().size() == 0
+    data.storage().resize_(size.numel())  # type: ignore[attr-defined]
```

Comment:
Curious, any reason these are `name_` instead of `_name`? And any reason that we want to expose these APIs to users?

Reply:
it is private...? hmm, let me make the naming convention to be consistent